zerver/models.py

```diff
@@ -373,10 +375,40 @@ def flush_realm_emoji(sender, **kwargs):
 post_save.connect(flush_realm_emoji, sender=RealmEmoji)
 post_delete.connect(flush_realm_emoji, sender=RealmEmoji)
 
-class RealmFilter(ModelReprMixin, models.Model):
+def filter_pattern_validator(value):
+    # type: (text_type) -> None
+    if not value:
+        raise ValidationError('Filter pattern cannot be empty')
+
+    prefix = value[0]
+    pattern = str(value[1:])
+    regex = re.compile(r'(\(\?P<\w+>.+\))')
+
+    if prefix not in settings.ALLOWED_FILTER_PREFIXES:
+        raise ValidationError(
+            'Filter pattern cannot start with `%s`, use one of the valid prefixes: %s' %
+            (value[0], ', '.join(settings.ALLOWED_FILTER_PREFIXES)))
+
+    try:
+        re.compile(pattern)
+    except:
+        # Regex is invalid
+        raise ValidationError('Invalid filter pattern')
+
+    if not regex.match(pattern):
+        raise ValidationError('Pattern format string must be in the following format: (?P<\w+>.+)')
+
+def filter_format_validator(value):
+    # type: (str) -> None
+    regex = re.compile(r'.*(%\(\w+\))')
```

Comment:
this regex seems like it could use some work to check the entire string to be of the format suggested in the error message below.  I think basically we want to ensure that it contains no regex special characters other than a single instance of `%([a-zA-Z0-9_0]+)s`.